docs/source/debugging.rst

```diff
@@ -0,0 +1,196 @@
+
+Exceptions
+----------
+
+When a task in your computation fails the standard way of understanding what
+went wrong is to look at the exception and traceback.  Often people do this
+with the ``pdb`` module, IPython ``%debug`` or ``%pdb`` magics, or by just
+looking at the traceback and investigating where in their code the exception
+occurred.
+
+Normally when a computation computes in a separate thread or a different
+machine these approaches break down.  Dask provides a few mechanisms to
+recreate the normal Python debugging experience.
+
+Inspect Exceptions and Tracebacks
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+By default, Dask already copies the exception and traceback wherever they
+occur and reraises that exception locally.  If your task failed with a
+``ZeroDivisionError`` remotely then you'll get a ``ZeroDivisionError`` in your
+interactive session.  Similarly you'll see a full traceback of where this error
+occurred, which, just like in normal Python, can help you to identify the
+troublsome spot in your code.
+
+However, you can not use the ``pdb`` module or ``%debug`` IPython magics with
+these tracebacks to look at the value of variables during failure.  You can
+only inspect things visually.  Additionally, the top of the traceback may be
+filled with functions that are dask-specific and not relevant to your
+problem, you can safely ignore these.
+
+Both the single-machine and distributed schedulers do this.
+
+
+Use the Single-Threaded Scheduler
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Dask ships with a simple single-threaded scheduler.  This doesn't offer any
+parallel performance improvements, but does run your Dask computation
+faithfully in your local thread, allowing you to use normal tools like ``pdb``,
+``%debug`` IPython magics, the profiling tools like the ``cProfile`` module and
+`snakeviz <https://jiffyclub.github.io/snakeviz/>`_.  This allows you to use
+all of your normal Python debugging tricks in Dask computations, as long as you
+don't need parallelism.
+
+This only works for single-machine schedulers.  It does not work with
+dask.distributed unless you use comfortable using the Tornado API (want to look
+at the `testing infrastructure
+<http://distributed.readthedocs.io/en/latest/develop.html#writing-tests>`_
+docs, which accomplish this).
+
+
+Rerun Failed Task Locally
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+If a remote task fails, we can collect the function and all inputs, bring them
+to the local thread, and then rerun the function in hopes of triggering the
+same exception locally, where normal debugging tools can be used.
+
+With the single machine schedulers, use the ``rerun_exceptions_locally=True``
+keyword.
+
+.. code-block:: python
+
+   x.compute(rerun_exceptions_locally=True)
+
+On the distributed scheduler use the ``recreate_error_locally`` method on
+anything that contains ``Futures`` :
+
+.. code-block:: python
+
+   >>> x.compute()
+   ZeroDivisionError(...)
+
+   >>> %pdb
+   >>> client.recreate_error_locally(x)
```

Comment:
Yes, something like that.  The intent seems the same to me even if the internals are necessarily different.  Is there particular confusion that you expect?

We probably can't change the dask/dask name at this stage (it has been around for a while) but could change the dask/distributed name.